### PR TITLE
[breaking] remove support for encoded directories

### PR DESCRIPTION
**Pull request by benmccann**
This is technically breaking, but only barely. Encoded directories mostly didn't work save for 11 characters. It's not clear to me it's possible to support more than that and those characters can generally be represented in a much nicer way in decoded format anyway. There's really no reason a user should have been writing `%40` instead of `@` in their directory name in the first place. I think `#`, `?` (only on Linux and Mac), and `:` (only on certain Windows systems) are the only ones that might be a bit less convenient to write as you'd now have to use a parameter matcher now. This seems like a good trade-off for making the code and docs much easier to understand

I removed the `$SVLT` test simply because it's already covered by other tests such as the `@svelte` test. The others were removed because they were testing encoded characters in directories which we no longer support

Closes https://github.com/sveltejs/kit/issues/7554
Closes https://github.com/sveltejs/kit/issues/7567
Closes https://github.com/sveltejs/kit/pull/7573
Closes https://github.com/sveltejs/kit/pull/7563

Probably solves https://github.com/sveltejs/kit/issues/7577 although I'm going to leave that open for a bit longer as I'd like to check on a couple more things first

### CODE REVIEW

**Comment by Conduitry** [comment]
We initially had this because we wanted to be able to support routes with arbitrary special characters in them. Without this, there's no way of having a route with literal characters in it that would normally have special meaning to SK other than to create a dummy param with a matcher that matches that one specific string. This works, but is ugly and wasteful. I don't think we want to completely remove support for encoded directory names.

**Comment by benmccann** [comment]
Yes, however, that has never actually worked. I tried dozens of versions going back more than a year to at least Vite 2.5 and none of them supported putting `%5b` (an encoded `[`) in the route except as a parameter. So this doesn't change anything in that regard. And more importantly, as far as I can tell, I don't think it's something Vite supports and thus would be impossible for us to implement even if we wanted to.

If we make this change now it puts us in a safe place for 1.0 where what we claim to support and actually support match. It would be backwards compatible to add support for encoded routes in the future if anyone else can figure out how or there's a new release of Vite that begins supporting it. In the meantime, it's such an uncommon character to have in a URL and no one has ever filed an issue about it, so I don't think it's a very common use case that we need to implement before 1.0.

**Comment by dummdidumm** [comment]
I haven't followed the (de)coding madness in the past, so how big of a breaking change is this? Were people previously forced to do `%40[username]` and now they are forced to do `@[username]`, or did `@[username]` always work, so people had two ways of doing the same? How often do people want to have `#` or `?` in their filenames (any examples of that in the wild)? Does this have any effect on the characters encoded _inside_ parameters (`[blog]` where `blog` is `does-this-work%3F`; `%3F` is `?`)?

**Comment by benmccann** [comment]
`@` always worked, so this is a very minor breaking change. There are only three characters that are somewhat affected per the PR description. I don't believe you were ever allowed a `%` inside brackets as part of a parameter name

**Comment by dummdidumm** [comment]
I mean the potential impact to people who have written down their paths in an encoded manner at the moment (`hello/world%3F/+page.svelte`) - from what I understand, this would no longer be possible, which might be a bigger breaking change than we think.

**Comment by benmccann** [comment]
There were only 11 characters that were even possible to write in encoded format previously. I've honestly never seen one in the wild. `hello/world%3F` is a pretty ugly URL, so people don't tend to hard code symbols like that in URLs but only use them in dynamic contexts.

File: packages/kit/src/utils/routing.js [hunk 1]
```diff
@@ -21,17 +21,26 @@ export function parse_route_id(id) {
...
+							if (segment.includes('%')) {
+								throw new Error(`Routes ${id} must be provided in decoded form.`);
+							}
+							if (/[#\/\?]/.test(segment)) {
+								// #, ?, and / will not be decoded by decodeURI so cannot be matched
+								// '/' is illegal on all file systems so we don't bother warning about it
+								throw new Error(
+									`Route ${id} should not contain '#' or '?'. Please use a parameter.`
+								);
+							}
```
Comment (Rich-Harris) [inline packages/kit/src/utils/routing.js:33]: this should happen sooner, in `create_manifest_data`. if it's here then this code will be sent to the client, needlessly

**Comment by Rich-Harris** [comment]
Been noodling around on this stuff locally and while I was able to get everything basically working for SSR and in a built app, it fails when JS kicks in because Vite can't load modules with certain characters in them. I think it's probably a Vite bug, but I also think that using URL-encoded characters in filenames is always going to be fraught.

So I agree that we should remove support for URL-encoded characters in directory names. I _don't_ agree that we should prevent people from using those characters in their routes though. Asking people to do `foo[x=colon]bar` for `/foo:bar` — creating a useless `x` param and forcing the creation of a `colon` matcher — is user-hostile in my view.

So we need a way of encoding certain characters that Vite can understand and that works on all filesystems. Ideally, it would be familiar to web developers. Luckily, just such a method exists: HTML entities. The example above would become `foo&colon;bar`.

Ideally, the decoding would happen as early in the process as possible. I think we should be able to decode everything except the characters that that are meaningful to SvelteKit: `[ ] ( )`. (Aside: we've talked about this before without resolution — do we need to reserve any other characters, in case we want to implement something like Next's proposed [dynamic parallel routes](https://nextjs.org/blog/layouts-rfc#dynamic-parallel-routes)? Or are we okay with something like that requiring a major version bump?)

**Pull request closed without merging**